### PR TITLE
Introduce a "stamp" IteratorResult

### PR DESCRIPTION
**Pull request by defunctzombie**
**User-Facing Changes**
Reduce the number of network requests when subscribed to sparsely populated topics and paused.

**Description**
The BufferedIterableSource is responsible for keeping a time-based read-ahead buffer of messages for playback. It does this using a producer and consumer pattern where the consumer reads from the buffer and updates the `readHead` location while the producer keeps messages buffered based on the read head. To keep the buffer filled, the producer reads messages from the underlying data source message iterator until some `readUntil` time which is a fixed offset ahead of the `readHead`. The producer stops when it encounters a message occurring after the `readUntil` time and waits for the consumer to move the `readHead` forward before attempting to read more messages.

The DataPlatformIterableSource implements two modes for message iteration. `partial` and `full`. In `partial` mode (the mode used for playback buffering), the implementation reads messages from the API using fixed-time request windows (currently 5 seconds). This is because we cannot control back-pressure on the request and want to avoid downloading large quantities of data prior to the user needing to play this data.

When opening a data platform stream and subscribing to a sparsely populated topic, the above behaviors result in many (hundreds to thousands) of requests to fetch each next 5 seconds of data from the api. This happens because the buffered source continues to read from the underlying source (data platform) until it can encounter a message _after_ the `readUntil` time. For a sparsely populated dataset it could be minutes or hours before the next such message exists and we get there in 5 second request increments.

This change provides a workaround to this behavior by introducing a new type of IteratorResult. Unlike the `messageEvent` or `problem` IteratorResult, the _stamp_ iterator result contains only a timestamp (no message or problem). The underlying source uses the stamp result to indicate it has loaded data through the stamp time as a signal to upstream message iterator consumers. The BufferedIterableSource is updated with logic to check the `stamp` time against the `readUntil` time to identify if the underlying source has been sufficiently buffered for playback. The DataPlatformIteraboeSource is updated to emit `stamp` results after each 5 second stream ends. Even if there are no messages in the stream, the stamp result indicates to callers that the specific timestamp has been reached.

Fixes: #4744

### CODE REVIEW

File: packages/studio-base/src/players/IterablePlayer/IIterableSource.ts [hunk 1]
```diff
@@ -56,14 +56,18 @@ export type MessageIteratorArgs = {
...
   | {
-      connectionId: number | undefined;
-      msgEvent: undefined;
+      type: "problem";
+      connectionId?: number;
       problem: PlayerProblem;
+    }
+  | {
+      type: "stamp";
+      stamp: Time;
```
Comment (jtbandes) [inline packages/studio-base/src/players/IterablePlayer/IIterableSource.ts:70]: Need comment on what this means. What does the time represent?

File: packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts [hunk 2]
```diff
@@ -776,13 +785,32 @@ export class IterablePlayer implements Player {
     const targetTime = add(this._currentTime, fromMillis(rangeMillis));
     const end: Time = clampTime(targetTime, this._start, this._untilTime ?? this._end);
 
+    if (this._lastStamp) {
+      // If the last message we saw is still ahead of the tick end time, we don't emit anything
+      if (compare(this._lastStamp, end) >= 0) {
```
Comment (jtbandes) [inline packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts:790]: Does `_lastStamp` correspond to the "last message we saw"? I thought a stamp result was different from a message result.

**Review by jtbandes — APPROVED** [review]
Gave the player stuff a cursory read. Can't say I fully understand it but am glad we have new tests being added 👍

File: packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts [hunk 3]
```diff
@@ -776,13 +785,32 @@ export class IterablePlayer implements Player {
...
+        this._messages = [];
+        this._queueEmitState();
+
+        if (this._untilTime && compare(this._currentTime, this._untilTime) >= 0) {
+          this.pausePlayback();
+        }
+        return;
+      }
+
+      this._lastStamp = undefined;
```
Comment (snosenzo) [inline packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts:804]: this whole sequence is a little unclear to me, could use a few more comments. I just don't understand why these things are happening outside of not emiting state when the lastStamped > endTIme

File: packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts (hunk 2 above)
Comment (defunctzombie) [reply packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts]: I've removed this comment and added a new one above the parent `if` statement.

File: packages/studio-base/src/players/IterablePlayer/IIterableSource.ts (hunk 1 above)
Comment (defunctzombie) [reply packages/studio-base/src/players/IterablePlayer/IIterableSource.ts:82]: Added a comment for the `IteratorResult` type.

File: packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts (hunk 3 above)
Comment (defunctzombie) [reply packages/studio-base/src/players/IterablePlayer/IterablePlayer.ts:810]: I've added a comment to the parent `if` statement describing the logic and why this bail-out of the tick exists.

**Pull request merged**